### ISSUE TITLE
Value Error while creating a Statistical Shape Model

### ISSUE THREAD

**Issue by LucaLumetti** [morphomatics/morphomatics#3]
I'm trying to create a SSM starting from 3D volumes.
My 3D volumes are numpy arrays obtained from CT scans and 1s are present where the object has been detected, 0 otherwise.

As far as I understood, I need to convert such np.array into a Subject. To perform this task I've written the following code:
```
import pyvista as pv

for f in files:
  volume = np.load(f"./{f}.npy")
  volume = volume.nonzero()
  volume = np.transpose(volume)
  cloud = pv.PolyData(volume)
  mesh = cloud.delaunay_3d(alpha=2.)
  mesh = mesh.extract_geometry()
  surf = Surface(mesh.points, mesh.faces.reshape(-1, 4)[:, 1:])
  surfaces.append(surf)
```
Until here everything works fine, I can also plot the mesh with `mesh.plot(notebook=True)` and it looks correct.
When I try to compute the SSM with:
```
SSM = StatisticalShapeModel(lambda ref: FundamentalCoords(ref))
SSM.construct(surfaces)
```
I get the error reported below, which is caused by calling `FundamentalCoords()` on the first surface. Even by trying different surfaces i still get the same error.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-12-eeca11b498b2>](https://localhost:8080/#) in <module>()
      8 # construct model
      9 SSM = StatisticalShapeModel(lambda ref: FundamentalCoords(ref))
---> 10 SSM.construct(surfaces)

8 frames
[/usr/local/lib/python3.7/dist-packages/morphomatics/stats/StatisticalShapeModel.py](https://localhost:8080/#) in construct(self, surfaces)
     47 
     48         # compute mean / setup manifold space (s.t. mean agrees to reference)
---> 49         self.setupShapeSpace(surfaces)
     50         coords = [self.space.to_coords(S.v) for S in surfaces]
     51         self.mean_coords = Mean.compute(self.space, coords)

[/usr/local/lib/python3.7/dist-packages/morphomatics/stats/StatisticalShapeModel.py](https://localhost:8080/#) in setupShapeSpace(self, surfaces, max_outer, max_inner)
     65         # initial guess
     66         mean = surfaces[0].copy()
---> 67         space:ShapeSpace = self.shape_space_cstr(mean)
     68 
     69         # iterate until mean and reference agree

[<ipython-input-12-eeca11b498b2>](https://localhost:8080/#) in <lambda>(ref)
      7 #FundamentalCoords(surfaces[0])
      8 # construct model
----> 9 SSM = StatisticalShapeModel(lambda ref: FundamentalCoords(ref))
     10 SSM.construct(surfaces)

[/usr/local/lib/python3.7/dist-packages/morphomatics/manifold/FundamentalCoords.py](https://localhost:8080/#) in __init__(self, reference, structure, metric_weights)
     54         self.metric_weights = (omega_C, omega_U)
     55 
---> 56         self.spanning_tree_path = self.setup_spanning_tree_path()
     57 
     58         # rotation and stretch manifolds

[/usr/local/lib/python3.7/dist-packages/morphomatics/manifold/FundamentalCoords.py](https://localhost:8080/#) in setup_spanning_tree_path(self)
    509             idx = idcs.pop(0)
    510             d = depth[idx] + 1
--> 511             neighs = self.ref.inner_edges.getrow(idx).tocoo()
    512 
    513             for neigh, edge in zip(neighs.col, neighs.data):

[/usr/local/lib/python3.7/dist-packages/morphomatics/geom/misc.py](https://localhost:8080/#) in memofn(self, *args, **kwargs)
     26             cache = getattr(self, cache_name, None)
     27             if id(fn) not in cache:
---> 28                 cache[id(fn)] = fn(self)
     29             return cache[id(fn)]
     30 

[/usr/local/lib/python3.7/dist-packages/morphomatics/geom/surface.py](https://localhost:8080/#) in inner_edges(self)
    117 
    118         # k-th (inner) edge, s.t. FFU(edgesI[k], edgesJ[k]) -> FFU(edgesJ[k], edgesI[k])
--> 119         tris2InnerEdgeId = sparse.coo_matrix((np.arange(1, FFU.getnnz()+1), (edgeTriI[:], edgeTriJ[:])), (m, m))
    120         tris2InnerEdgeId = tris2InnerEdgeId + tris2InnerEdgeId.T
    121         tris2InnerEdgeId.data = tris2InnerEdgeId.data -1

[/usr/local/lib/python3.7/dist-packages/scipy/sparse/coo.py](https://localhost:8080/#) in __init__(self, arg1, shape, dtype, copy)
    194             self.data = self.data.astype(dtype, copy=False)
    195 
--> 196         self._check()
    197 
    198     def reshape(self, *args, **kwargs):

[/usr/local/lib/python3.7/dist-packages/scipy/sparse/coo.py](https://localhost:8080/#) in _check(self)
    281         if self.nnz > 0:
    282             if self.row.max() >= self.shape[0]:
--> 283                 raise ValueError('row index exceeds matrix dimensions')
    284             if self.col.max() >= self.shape[1]:
    285                 raise ValueError('column index exceeds matrix dimensions')

ValueError: row index exceeds matrix dimensions
```

**Comment by ambellan** [comment]
Dear Luca, be warmly welcomed to the world of morphomatics. There two issues I would like to comment on and probably this will help you to find out what went wrong and how to fix it.

1. The error you are facing occurs pretty much in the beginning of the shape model construction. The first input surface is taken and as it does not exist yet, a spanning tree path along all inner edges of that surface is constructed (`self.setup_spanning_tree_path()`). This tree is needed to properly initialize the shape reconstruction problem from a point in shape space to a surface in 3d space. Here the error message states that something went wrong when determining the inner edges (`self.ref.inner_edges.getrow(idx).tocoo()`) and this can have different reasons. Please check that: (a) Your surface is oriented, i.e. the numbering of points forming triangles is consistent for all triangles (equivalently all triangle normals point in the same direction [outside/inside]). (b) Your surface is manifold, i.e. there is no edge shared by more than two triangles. (c) Your surface is connected. This is assumption is needed again for the reconstruction problem as the underlying shape space does not care about relative positioning of connected components.

2. When employing morphomatics, your input shapes must be in correspondence (vertex-wise and triangle-wise). This directly implies that all your shapes have to feature a fixed topology. However, I am not sure whether your current workflow can guarantee correspondence. You probably have to establish this correspondence right after `mesh = mesh.extract_geometry()`. This can be achieved in various different ways, you could e.g. solve the non-rigid registration problem by applying gaussian regularized [Coherent Point Drift](https://github.com/siavashk/pycpd) or the diffeomorphism based [Deformetrica](https://pypi.org/project/deformetrica/) setup.

**Comment by LucaLumetti** [comment]
Hi Felix, thanks for the really in-depth explanation.
I figured out that all the requirements you stated in point 1 weren't satisfied and managed to overcome them by setting a different `alpha` in `delaunay_3d`.
About point 2, I'm not exactly sure what 

> your input shapes must be in correspondence (vertex-wise and triangle-wise)

 means. As far as I know and understand by looking at the link you provided me, solving the non-rigid registration problem means finding a matrix (or a kernel and a matrix) that transforms a source shape into a given target shape. I was also able to find such a matrix by using the Coherent Point Drift algorithm, but I don't get how I can use that to make a shape in correspondence with another.
I think that his lack of correspondence is the cause of the error I'm getting now (I paste it below just for completion)

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-5-eeca11b498b2>](https://localhost:8080/#) in <module>()
      8 # construct model
      9 SSM = StatisticalShapeModel(lambda ref: FundamentalCoords(ref))
---> 10 SSM.construct(surfaces)

5 frames
[/usr/local/lib/python3.7/dist-packages/morphomatics/stats/StatisticalShapeModel.py](https://localhost:8080/#) in construct(self, surfaces)
     47 
     48         # compute mean / setup manifold space (s.t. mean agrees to reference)
---> 49         self.setupShapeSpace(surfaces)
     50         coords = [self.space.to_coords(S.v) for S in surfaces]
     51         self.mean_coords = Mean.compute(self.space, coords)

[/usr/local/lib/python3.7/dist-packages/morphomatics/stats/StatisticalShapeModel.py](https://localhost:8080/#) in setupShapeSpace(self, surfaces, max_outer, max_inner)
     72         for _ in range(max_outer):
     73             # compute mean in manifold space
---> 74             x = Mean.compute(space, [space.to_coords(S.v) for S in surfaces], x=space.ref_coords if len(surfaces) > 2 else None, max_iter=max_inner)
     75 
     76             # compute vertex coordinates of mean

[/usr/local/lib/python3.7/dist-packages/morphomatics/stats/StatisticalShapeModel.py](https://localhost:8080/#) in <listcomp>(.0)
     72         for _ in range(max_outer):
     73             # compute mean in manifold space
---> 74             x = Mean.compute(space, [space.to_coords(S.v) for S in surfaces], x=space.ref_coords if len(surfaces) > 2 else None, max_iter=max_inner)
     75 
     76             # compute vertex coordinates of mean

[/usr/local/lib/python3.7/dist-packages/morphomatics/manifold/FundamentalCoords.py](https://localhost:8080/#) in to_coords(self, v)
    114         """
    115         # compute gradients
--> 116         D = self.ref.grad @ v
    117 
    118         # decompose...

[/usr/local/lib/python3.7/dist-packages/scipy/sparse/base.py](https://localhost:8080/#) in __matmul__(self, other)
    558             raise ValueError("Scalar operands are not allowed, "
    559                              "use '*' instead")
--> 560         return self.__mul__(other)
    561 
    562     def __rmatmul__(self, other):

[/usr/local/lib/python3.7/dist-packages/scipy/sparse/base.py](https://localhost:8080/#) in __mul__(self, other)
    514 
    515             if other.shape[0] != self.shape[1]:
--> 516                 raise ValueError('dimension mismatch')
    517 
    518             result = self._mul_multivector(np.asarray(other))

ValueError: dimension mismatch
```

Thank you again for the help, I really appreciate it!

**Comment by vontycowicz** [comment]
Hi Luca, also from me a warm welcome.

Yes, the error you are seeing is surely because of the missing correspondence.
You might get a better understanding of what is expected by looking at the [Discrete representation](https://morphomatics.github.io/tutorials/tutorial_shape_space/#discrete-representation) section in our _shape space_ tutorial.
In particular, each ```Surface``` instance has to be created with the same array ```f``` of vertex indices forming the triangles. The array  of vertex coordinates ```v``` should hold displaced positions obtain, e.g., via CPD.

You can also take a look at the meshes used in the tutorials (see [here](https://github.com/morphomatics/morphomatics.github.io/tree/master/data) to get an idea.

**Comment by YuliangXiaoYLX** [comment]
@vontycowicz Hi, currently I used the ```PointDistributionModel``` to construct the SSM for my project. However, there is an error about the ```SSM.mean.v``` whose data type is ```class 'jaxlib.xla_extension.DeviceArray'``` instead of ```ndarray```. I checked the source code and found that the error may happen when call ```v = space.from_coords(x)``` where in the ```PointDistributionModel``` it used ```c.reshape(self.ref.v.shape)``` which did not convert from ```vectorized point distribution coords``` to ```vertex coords```

**Comment by vontycowicz** [comment]
Dear Yuliang,

thanks for your interest in Morphomatics and spotting the issue.
I had a look at `PointDistributionModel` and fixed some bugs: Check out the new 2.1.1 release.
That said, you might still encounter clashes between `jax` and `numpy` arrays as we are using both at different places. The former can be converted to the latter simply using `np.asarray`.

Take care,
Christoph

P.S.: Should you spot further glitches, consider opening a new issue such that its easier to find for other users as well.

**Comment by YuliangXiaoYLX** [comment]
> Dear Yuliang,
> 
> thanks for your interest in Morphomatics and spotting the issue. I had a look at `PointDistributionModel` and fixed some bugs: Check out the new 2.1.1 release. That said, you might still encounter clashes between `jax` and `numpy` arrays as we are using both at different places. The former can be converted to the latter simply using `np.asarray`.
> 
> Take care, Christoph
> 
> P.S.: Should you spot further glitches, consider opening a new issue such that its easier to find for other users as well.

Thanks for your reply, I will start a new issue for others
